librz/analysis/block.c

```diff
@@ -952,3 +952,169 @@ RZ_API void rz_analysis_block_automerge(RzList *blocks) {
+	if (analysis->opt.jmpmid && x86) {
+		BlockFromOffsetJmpmidCtx ctx = { off, NULL };
+		rz_analysis_blocks_foreach_in (analysis, off, block_from_offset_jmpmid_cb, &ctx);
+		return ctx.ret;
+	}
+
+	RzAnalysisBlock *ret = NULL;
+	rz_analysis_blocks_foreach_in (analysis, off, bb_from_offset_first_cb, &ret);
+	return ret;
+}
+
+/**
+ * @return the offset of the i-th instruction in the basicblock bb or U16_MAX if i is invalid.
+ */
+RZ_API ut16 rz_analysis_block_get_op_offset(RzAnalysisBlock *block, size_t i) {
+	if (i >= block->ninstr) {
+		return UT16_MAX;
+	}
+	return (i > 0 && (i - 1) < block->op_pos_size) ? block->op_pos[i - 1] : 0;
+}
+
+/**
+ * @return the absolute address of the i-th instruction in block or UT64_MAX if i is invalid.
+ */
+RZ_API ut64 rz_analysis_block_get_op_addr(RzAnalysisBlock *block, size_t i) {
+	ut16 offset = rz_analysis_block_get_op_offset (block, i);
+	if (offset == UT16_MAX) {
+		return UT64_MAX;
+	}
+	return block->addr + offset;
+}
+
+/**
+ * set the offset of the i-th instruction in the basicblock bb
+ */
+RZ_API bool rz_analysis_block_set_op_offset(RzAnalysisBlock *block, size_t i, ut16 v) {
+	// the offset 0 of the instruction 0 is not stored because always 0
+	if (i > 0 && v > 0) {
+		if (i >= block->op_pos_size) {
+			size_t new_pos_size = i * 2;
+			ut16 *tmp_op_pos = realloc (block->op_pos, new_pos_size * sizeof (*block->op_pos));
+			if (!tmp_op_pos) {
+				return false;
+			}
+			block->op_pos_size = new_pos_size;
+			block->op_pos = tmp_op_pos;
+		}
+		block->op_pos[i - 1] = v;
+		return true;
+	}
+	return true;
+}
+
+/**
+ * @return the address of the instruction that occupies a given offset or UT64_MAX if off is not in the block.
+ */
+RZ_API ut64 rz_analysis_block_get_op_addr_in(RzAnalysisBlock *bb, ut64 off) {
+	ut16 delta, delta_off, last_delta;
+	int i;
+
+	if (!rz_analysis_block_contains (bb, off)) {
+		return UT64_MAX;
+	}
+	last_delta = 0;
+	delta_off = off - bb->addr;
+	for (i = 0; i < bb->ninstr; i++) {
+		delta = rz_analysis_block_get_op_offset (bb, i);
+		if (delta > delta_off) {
+			return bb->addr + last_delta;
+		}
+		last_delta = delta;
+	}
+	return bb->addr + last_delta;
+}
+
+/**
+ * @return the size of the i-th instruction in a basic block
+ */
+RZ_API ut64 rz_analysis_block_get_op_size(RzAnalysisBlock *bb, size_t i) {
+	if (i >= bb->ninstr) {
+		return UT64_MAX;
+	}
+	ut16 idx_cur = rz_analysis_block_get_op_offset (bb, i);
+	ut16 idx_next = rz_analysis_block_get_op_offset (bb, i + 1);
+	return idx_next != UT16_MAX? idx_next - idx_cur: bb->size - idx_cur;
+}
+
+/**
+ * Successively disassemble the ops in this block, update the contained op addrs
+ * and create xrefs if necessary.
+ * This will not move or resize the block itself or touch anything else around it,
+ * it is primarily useful when creating or editing blocks after full function analysis.
+ */
+RZ_API void rz_analysis_block_analyze_ops(RzAnalysisBlock *block) {
+	rz_return_if_fail (block);
```

Comment:
should you check if blocksize is > 0 ?

Reply:
Not in the assert, but yeah that wouldn't hurt.